pkg/controllers/awsloadbalancercontroller/controller.go

```diff
@@ -134,25 +134,32 @@ func (r *AWSLoadBalancerControllerReconciler) Reconcile(ctx context.Context, req
 		}
 	}
 
-	var credentialsRequest *cco.CredentialsRequest
-	if credentialsRequest, err = r.ensureCredentialsRequest(ctx, r.Namespace, lbController); err != nil {
-		return ctrl.Result{}, fmt.Errorf("failed to ensure CredentialsRequest for AWSLoadBalancerController %q: %w", req.Name, err)
+	credSecretNsName := types.NamespacedName{Namespace: r.Namespace}
+	if lbController.Spec.Credentials == nil {
+		credentialsRequest, err := r.ensureCredentialsRequest(ctx, r.Namespace, lbController)
+		if err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed to ensure CredentialsRequest for AWSLoadBalancerController %q: %w", req.Name, err)
+		}
+		credSecretNsName.Name = credentialsRequest.Spec.SecretRef.Name
```

Comment:
> You should check whether credentialsRequest is nil, which could conceivably happen if ensureCredentialsRequest created the CredentialsRequest and then the API hiccuped and returned a not-found error when ensureCredentialsRequest tried to get the newly created CredentialsRequest.

Not sure I understood in which case `ensureCredentialsRequest` can return `nil` without an error.

> (By the way, it looks like ensureCredentialsRequest does an extra get if the CredentialsRequest already exists and doesn't need an update.)

True, the second get is unnecessary if no update was done. Will try to change it but if it'll be a big of a change will create a dedicated PR.

Reply:
> > You should check whether credentialsRequest is nil, which could conceivably happen if ensureCredentialsRequest created the CredentialsRequest and then the API hiccuped and returned a not-found error when ensureCredentialsRequest tried to get the newly created CredentialsRequest.
> 
> Not sure I understood in which case `ensureCredentialsRequest` can return `nil` without an error.

Theoretically, the following _could_ happen:

1. `ensureCredentialsRequest` calls `currentCredentialsRequest`, which does a get; API returns not-found, so `currentCredentialsRequest` returns false.  
2. `ensureCredentialsRequest` calls `createCredentialsRequest`, which creates the CredentialsRequest.  
3. `ensureCredentialsRequest` calls `currentCredentialsRequest` a second time, and the API again returns not-found, which shouldn't happen but theoretically could because of (* waves hands *) "eventual consistency" (e.g. maybe one control-plane node got the "create" request, the other one got the subsequent "get" request, and the two nodes are not quite synchronized—at least I think that's a possibility).  
4. `ensureCredentialsRequest` returns a nil pointer value and nil error value.  
5. `credSecretNsName.Name = credentialsRequest.Spec.SecretRef.Name` panics because `credentialsRequest` is nil.  

> > (By the way, it looks like ensureCredentialsRequest does an extra get if the CredentialsRequest already exists and doesn't need an update.)
> 
> True, the second get is unnecessary if no update was done. Will try to change it but if it'll be a big of a change will create a dedicated PR.

All right.  It isn't very important; just a small opportunity to improve efficiency.

Reply:
> All right. It isn't very important; just a small opportunity to improve efficiency.

Updated `ensureCredentialsRequest` to not do the second GET if the credentialsrequest wasn't updated.

I'm sorry I'm still not getting how it's possible to have nil credentialsrequest object without an errortaking into account that `NotFound` error is treated as error (not ignored).

Reply:
> I'm sorry I'm still not getting how it's possible to have nil credentialsrequest object without an error taking into account that `NotFound` error is treated as error (not ignored).

The last thing that `ensureCredentialsRequest` does is this:
```go
	_, current, err = r.currentCredentialsRequest(ctx, credReq)
	if err != nil {
		return nil, fmt.Errorf("failed to get existing credentials request %q: %w", credReq.Name, err)
	}
	return current, nil
```
`currentCredentialsRequest` does not treat `NotFound` as an error:
```go
		if errors.IsNotFound(err) {
			return false, nil, nil
```
So if `createCredentialsRequest` succeeds and then (for whatever reason) `currentCredentialsRequest` returns `false, nil, nil`, then `ensureCredentialsRequest` returns `nil, nil`.

Reply:
Sorry for the back and forth. You are right, I didn't dig enough. Also, `ensureCredentialsRequest` method name becomes misleading in that regard because it doesn't really ensure as long as there is this case.

I updated `ensureCredentialsRequest` to make sure `not found` is treated as error.